twitter-borrowbot/twitterbot.py

```diff
@@ -3,47 +3,69 @@
 import time
 import tweepy
 from dotenv import load_dotenv
+import traceback
 
 from services import InternetArchive, ISBNFinder, Logger
+from TwitterBotErrors import TweepyAuthenticationError, FileIOError, LastSeenIDError, GetMentionsError, TooManyMentionsError, FindISBNError, GetTweetError, GetEditionError, FindAvailableWorkError, SendTweetError
 
 ACTIONS = ('read', 'borrow', 'preview')
 READ_OPTIONS = dict(zip(InternetArchive.MODES, ACTIONS))
-BOT_NAME = "@borrowbot"
+BOT_NAME = None
+BOT_ID = None
 STATE_FILE = 'last_seen_id.txt'
+SLEEP_TIME = 15 # in seconds 
+MENTION_LIMIT = 128 
 
-LOGGER = Logger("./logs/tweet_logs.txt", "./logs/error_logs.txt")
+TWEET_LOGS = "./logs/tweet_logs.txt" 
+ERROR_LOGS = "./logs/error_logs.txt"
 
-load_dotenv()
+API = None
 
-# Authenticate
-auth = tweepy.OAuthHandler(
-    os.environ.get('CONSUMER_KEY'),
-    os.environ.get('CONSUMER_SECRET')
-)
-auth.set_access_token(
-    os.environ.get('ACCESS_TOKEN'),
-    os.environ.get('ACCESS_TOKEN_SECRET')
-)
-api = tweepy.API(auth, wait_on_rate_limit=True)
+def authenticate():
+    load_dotenv()
+    if not os.environ.get('CONSUMER_KEY') or not os.environ.get('CONSUMER_SECRET') or not os.environ.get('ACCESS_TOKEN') or not os.environ.get('ACCESS_TOKEN_SECRET'):
+        raise TweepyAuthenticationError(error="Missing .env file or missing necessary keys for authentication")
+    try:
+        # Authenticate
+        auth = tweepy.OAuthHandler(
+            os.environ.get('CONSUMER_KEY'),
+            os.environ.get('CONSUMER_SECRET')
+        )
+        auth.set_access_token(
+            os.environ.get('ACCESS_TOKEN'),
+            os.environ.get('ACCESS_TOKEN_SECRET')
+        )
+        global API
+        API = tweepy.API(auth, wait_on_rate_limit=True)
+        global BOT_NAME
+        BOT_NAME = '@' + API.me().screen_name # test authentication & set BOT_NAME
+        global BOT_ID
+        BOT_ID = API.me().id
```

Comment:
I notice you're trying to keep things isolated, but I think these abstractions may look more convoluted to me than the original code. Now we're locally setting globals and relying on the programmer to call `authenticate`.

Is there ever a case you can think of where we don't want `authenticate` called?

Nothing wrong with the error handling :)